File: 0337-house-robber-iii/0337-house-robber-iii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

// @Author: KING-SEN
// Programming Language Used: C++

class Solution {
public:
    pair<int, int> dfs(TreeNode* root){
        // Base case return as [rob, notRob] as pair answer
        if (root == NULL) {
            return {0, 0};
        }
        auto leftResult = dfs(root->left);
        auto rightResult = dfs(root->right);
        int rob = root->val + leftResult.second + rightResult.second;
        int notRob = max(leftResult.first, leftResult.second) + max(rightResult.first, rightResult.second);
        return {rob, notRob};
    }
    int rob(TreeNode* root) {
        pair<int, int> answer = dfs(root);
        return max(answer.first, answer.second);
    }
};


// Time Complexity: O(N), here N = Number of nodes in the binary tree
// Space Complexity: O(N), since we need stack to do recursion ans the maximum depth of the 
// recursion is the height of the tree, which is O(N) in the worst case and O(log N) in the best case

File: 0337-house-robber-iii/0337-house-robber-iii.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def dfs(self, root):
        # Base case return as [with Root, without Root] as pair answer
        if root is None:
            return [0, 0]
        left_result = self.dfs(root.left)
        right_result = self.dfs(root.right)
        with_root = root.val + left_result[1] + right_result[1]
        without_root = max(left_result) + max(right_result)
        return [with_root, without_root]
    
    def rob(self, root: Optional[TreeNode]) -> int:
        answer = self.dfs(root)
        return max(answer)  